src/main/kotlin/org/rust/ide/presentation/TypeRendering.kt

```diff
@@ -266,15 +257,30 @@ private data class TypeRenderer(
     private fun formatBoundElementGenerics(
         boundElement: BoundElement<RsGenericDeclaration>,
         render: (Ty) -> String
+    ): List<String> = formatGenerics(boundElement.element, boundElement.subst, render)
+
+    private fun formatGenerics(
+        declaration: RsGenericDeclaration,
+        subst: Substitution,
+        render: (Ty) -> String
     ): List<String> {
-        val tySubst = boundElement.element.typeParameters.map { render(boundElement.subst[it] ?: TyUnknown) }
-        val regionSubst = if (includeLifetimeArguments) {
-            boundElement.element.lifetimeParameters.map { render(boundElement.subst[it] ?: ReUnknown) }
-        } else {
-            emptyList()
-        }
-        val constSubst = boundElement.element.constParameters.map { render(boundElement.subst[it] ?: CtUnknown) }
-        return regionSubst + tySubst + constSubst
+        var nonDefaultTyFound = false
+        return declaration.genericParameters
+            .asReversed()
+            .filter { parameter ->
+                if (!skipUnchangedDefaultTypeArguments || parameter !is RsTypeParameter) return@filter true
+                nonDefaultTyFound = nonDefaultTyFound || parameter.typeReference?.type != subst[parameter]
```

Comment:
Is it possible to get `null` from `subst[parameter]` and `null` from `parameter.typeReference?.type`?
Maybe it's worth explicitly check that `parameter.typeReference != null`?